### PR TITLE
Split up `util.{hh,cc}`

### LINKED ISSUE #8913: Split `utils.hh`
`utils.hh` currently contains a mix of pure functions (e.g. `tokenizeString`) and functions to deal with the operating system (file descriptors, filesystem, signals etc.). It is also quite large.

It should be split up by topic, and downstream files should only `#include` the pats they need.

See also https://github.com/NixOS/nix/pull/8901. I already started splitting it up in that, to avoid a sea of CPP. But the right order in which to do things is to split it up first.

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Fixes #8913

All OS and IO operations should be moved out, leaving only some misc portable pure functions.

This is useful to avoid copious CPP when doing things like Windows and Emscripten ports.

# Context

This sort of thing is annoying to write and annoying to read, but I realized I have to do it before trying to finish #8901.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by fricklerhandwerk** [comment]
Triaged in Nix maintainers meeting:

- @Ericson2314: goal is to get closer to make Nix work on Windows
- @edolstra: would be better to make the changes when there is direct need
  - you will only know exactly what the ideal factoring is once you start the porting
- @tomberek: recently played around with providing an Emscripten build output, this would help with that kind of thing
- @edolstra: two more problems:
  - adds friction to using `git blame`
  - more compilation units increases compilation time
    - @Ericson2314: would rather focus on readability, can always add compilation performance hacks on top (e.g. pre-concatenating files)
      - can do follow up and also merge into lazy trees
 - idea approved

File: src/libutil/ambient-authority.hh
Comment (fricklerhandwerk) [inline src/libutil/ambient-authority.hh:1]: That's a weird name. What is it really about? Why is it not e.g. `system-specific.hh`?

File: src/libutil/file-descriptor.cc [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+void writeLine(int fd, std::string s)
+{
+    s += '\n';
+    writeFull(fd, s);
+}
+
+
+std::string drainFD(int fd, bool block, const size_t reserveSize)
+{
+    // the parser needs two extra bytes to append terminating characters, other users will
```
Comment (fricklerhandwerk) [inline src/libutil/file-descriptor.cc:80]: Which parser?

**Review by fricklerhandwerk — APPROVED** [review]
Okay, I gave it an almost complete but still cursory look. I think the desired end result is very desirable.

I fully trust you on doing the right thing here, so that's approved.

Yet I'm quite concerned about this being a *huge* diff that could allow mistakes to slip through into foundational components, and I wouldn't want to share responsibility for that. I'd feel a lot more comfortable with doing that in multiple steps, even if that ends up being 30% more work. It will also ease dissection a lot.

Multiple commits may be fine, but reviews leading to rewriting history will just produce friction as those tend to confuse me, so I'd err towards multiple PRs. If anything, it will make a lot more likely for me to really go through each component and make sure I understand it and its role in the system, so I see it as a good opportunity for knowledge sharing. Being involved in way too many things, spending hours in a row staring at code is something I simply can't afford these days.

If anyone else on the @nixos/nix-team says "nah it'll be fine" we can of course just merge it as is.

File: src/libutil/file-descriptor.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libutil/file-descriptor.cc:80]: I just moved that code :)

File: src/libutil/ambient-authority.hh
Comment (Ericson2314) [reply src/libutil/ambient-authority.hh:1]: I went with `current-process.hh`, with the new first commit `nativeSystem` is no longer there, and everything else is squarely about the current process.

**Comment by Ericson2314** [comment]
Thanks @fricklerhandwerk. So one that makes me feel a lot better about this is `git show --color-moved`. That "proves" that this is just moving code around, and thus while it's a lot of code motion and the drudgery of cutting and pasting was mind-melting, there isn't anything interesting going on.

I do agree knowledge sharing is good, but I an easier way to do that is going through these files *post split*, e.g. during the Windows porting process. It's much easier to digest things when they are already cut into bite sized pieces :).

As you saying trying to redo this into separate PRs for each header would take longer. While I don't really mind it for the headers *themselves*, it would worse code churn on the `#includes` in all the *other* files and headers came and went. Because of that, because we have *so many more interesting things* to spend reviewer time on (this is a boring step towards an interesting destination, Windows), and because you are on board with breaking up the giant file, I think it is just best to merge this as-is.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
These moves were long overdue. Merging while the merge base is up to date.

Not expecting much merge conflicts from this, as this is support code - not a place where features happen. Certainly we have better places than a "utils" file by now!

> `git show --color-moved`.

This is a life saver for review.

**Pull request merged**